Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp

```diff
@@ -1353,16 +1376,57 @@ void JitArm64::srawx(UGeckoInstruction inst)
       ComputeRC0(gpr.GetImm(a));
     return;
   }
-
-  if (gpr.IsImm(b) && !js.op->wantsCA)
+  else if (gpr.IsImm(s) && gpr.GetImm(s) == 0)
+  {
+    gpr.SetImmediate(a, 0);
+    ComputeCarry(false);
+    if (inst.Rc)
+      ComputeRC0(0);
+    return;
+  }
+  else if (gpr.IsImm(b))
   {
     int amount = gpr.GetImm(b);
-    if (amount & 0x20)
-      amount = 0x1F;
+
+    if (amount == 0)
+    {
+      ComputeCarry(false);
```

Comment:
1. By checking for zero before applying any masking, we may miss out on some cases where the entire register may not be zero, but the bits that matter for the shift are. Probably doesn't happen too often in practice, though.
2. Don't we at least need to do a MOV from gpr.R(s) to gpr.R(a) in the shift-by-zero case? (Disclaimer, not terribly familiar with ARM64...)

Reply:
> Don't we at least need to do a MOV from gpr.R(s) to gpr.R(a) in the shift-by-zero case? (Disclaimer, not terribly familiar with ARM64...)

Oops, yes :D

Reply:
> By checking for zero before applying any masking, we may miss out on some cases where the entire register may not be zero, but the bits that matter for the shift are. Probably doesn't happen too often in practice, though.

By masking, you mean masking out the top bit of `amount`, right? If so, it doesn't seem correct to me to do this check after masking.

Reply:
> By masking, you mean masking out the top bit of `amount`, right? If so, it doesn't seem correct to me to do this check after masking.

Yes.

https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/PowerPC/Interpreter/Interpreter_Integer.cpp#L310

Only the 6 lowest bits of the register matter, the rest can be ignored.

Reply:
Ah, right, this isn't about doing the `amount & 0x1f` masking, but rather the `amount & 0x3f` masking. I see now. Thanks!